plutus-core/plutus-core/src/PlutusCore/Builtin/Case.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes        #-}
+{-# LANGUAGE TypeOperators     #-}
+
+module PlutusCore.Builtin.Case where
+
+import PlutusCore.Core.Type (Type, UniOf)
+import PlutusCore.Name.Unique
+
+import Control.DeepSeq (NFData (..), rwhnf)
+import Data.Default.Class (Default (..))
+import Data.Text (Text)
+import Data.Vector (Vector)
+import NoThunks.Class
+import Text.PrettyBy (display)
+import Universe
+
+class AnnotateCaseBuiltin uni where
+    -- | Given a tag for a built-in type and a list of branches, annotate each of the branches with
+    -- its expected argument types or fail if casing on values of the built-in type isn't supported.
+    annotateCaseBuiltin
+        :: UniOf term ~ uni
+        => SomeTypeIn uni
+        -> [term]
+        -> Either Text [(term, [Type TyName uni ann])]
+
+class CaseBuiltin uni where
+    -- | Given a constant with its type tag and a vector of branches, choose the appropriate branch
+    -- or fail if the constant doesn't correspond to any of the branches (or casing on constants of
+    -- this type isn't supported at all).
+    caseBuiltin :: UniOf term ~ uni => Some (ValueOf uni) -> Vector term -> Either Text term
+
+data CaserBuiltin uni = CaserBuiltin
```

Comment:
It seems `CaserBuiltin` should render the `CaseBuiltin` class superfluous, since an instance of `CaseBuiltin` can just be represented with a `CaserBuiltin` value?

Also, why not newtype?

Reply:
> We do use CaseBuiltin as an actual constraint

Ok, so `Case` is used to define how a universe normally handles casing on builtins, and `Caser` exists so that it can vary based on protocol version. I guess that's fine, but not easy to tell, and their similar names don't help.

> See [this](https://github.com/IntersectMBO/plutus/blob/5f3483866477eee84e04ccc81147a39eacdddc47/plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs#L64-L71) for a similar case.

Create a Note and link to it?